### PR TITLE
logs update

### PR DESCRIPTION
**Pull request by SargtLa**

### CODE REVIEW

File: api/avatar.go [hunk 1]
```diff
@@ -83,7 +83,7 @@ func HandleAddAvatar(ctx *fasthttp.RequestCtx) (interface{}, error) {
 	}
 
 	if i > 0 {
-		toLogCandidate(ctx, DB, id, "avatar", CODE_LOG_UPDATE)
+		toLogCandidate(ctx, DB, id, "avatar", db.GetLogUpdateId())
```
Comment (ruslanBik4) [inline api/avatar.go]: I think that we need several functions which must doing different log operations.
toLogCandidateUpdate
toLogCandidateAdd
& others
 Th function must use constant within its body

File: api/candidate.go [hunk 2]
```diff
@@ -656,10 +662,24 @@ func HandleEditCandidate(ctx *fasthttp.RequestCtx) (interface{}, error) {
 			if i > 0 {
 				text += ", "
 			}
+			switch col {
```
Comment (ruslanBik4) [inline api/candidate.go]: must save changes as object jsonb

File: api/company.go [hunk 3]
```diff
@@ -62,7 +59,7 @@ func HandleAddCompany(ctx *fasthttp.RequestCtx) (interface{}, error) {
 		return createErrResult(err)
 	}
 
-	toLogCompany(ctx, DB, int32(id), "", CODE_LOG_INSERT)
+	toLogCompany(ctx, DB, int32(id), "", db.GetLogInsertId())
```
Comment (ruslanBik4) [inline api/company.go]: there ,are function like as logCandidate

File: api/consts.go [hunk 4]
```diff
@@ -12,15 +12,19 @@ const (
...
 
 const (
-	CODE_LOG_UPDATE     = 100
-	CODE_LOG_INSERT     = 101
-	CODE_LOG_PEFORM     = 102
-	CODE_LOG_DELETE     = 103
-	CODE_LOG_RE_CONTACT = 104
-	CODE_ADD_COMMENT    = 104
-	CODE_DEL_COMMENT    = 104
+	CODE_LOG_UPDATE        = 100
```
Comment (ruslanBik4) [inline api/consts.go]: rm here & get from database

File: cfg/DB/func/get_log.ddl [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+       CONCAT('Рекрутер ',
+              users.name,
+              log_actions.text_before_cand,
+              CASE
+                  WHEN logs.candidate_id > 0 THEN CONCAT(log_actions.for_candidate, can.name)
+                  ELSE CONCAT(log_actions.for_company, companies.name) END,
+              log_actions.text_after_cand,
+              CASE
+                  WHEN log_actions.name = 'CODE_SEND_CV' or log_actions.name = 'CODE_APPOINT_INTERVIEW'
+                      THEN CONCAT(' на вакансию ', platforms.name, ', ', seniorities.name,
```
Comment (ruslanBik4) [inline cfg/DB/func/get_log.ddl:22]: use Format function

File: cfg/DB/func/get_log.ddl [hunk 6]
```diff
@@ -0,0 +1,44 @@
+CREATE OR REPLACE FUNCTION get_log_test(Id integer, isCand bool)
+    RETURNS table(
```
Comment (ruslanBik4) [inline cfg/DB/func/get_log.ddl:2]: will be better use json output

File: cfg/DB/func/get_log.ddl [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+       logs.create_at as date
+
+from logs left Join companies on (logs.company_id = companies.id)
+    left join vacancies ON (logs.vacancy_id = vacancies.id)
+    join users ON (logs.user_id = users.id)
+    join candidates can ON (logs.candidate_id = can.id)
+    left Join platforms ON (vacancies.platform_id = platforms.id)
+    left Join seniorities ON (vacancies.seniority_id = seniorities.id)
+    left Join log_actions ON (logs.action_code = log_actions.id)
+where (logs.candidate_id = $1 AND $2 = true) or (logs.company_id = $1 AND $2 = false)
```
Comment (ruslanBik4) [inline cfg/DB/func/get_log.ddl]: for boolean value not need use '= true'

File: cfg/DB/table/logs.ddl [hunk 8]
```diff
@@ -27,4 +27,12 @@ alter table logs
 -- alter table logs
 --     add constraint logs_vacancy_id_fk
 --         foreign key (vacancy_id) references vacancies
---             on update cascade on delete cascade;
\ No newline at end of file
+--             on update cascade on delete cascade;
+
+update logs
+set action_code=108
+where logs.text LIKE '% назначил %';
```
Comment (ruslanBik4) [inline cfg/DB/table/logs.ddl:34]: why not accord with 'kod_deystviya'?

**Review by ruslanBik4 — CHANGES_REQUESTED** [review]

File: cfg/DB/func/get_log.ddl (hunk 5 above)
Comment (SargtLa) [reply cfg/DB/func/get_log.ddl:22]: don't get it

File: cfg/DB/func/get_log.ddl (hunk 6 above)
Comment (SargtLa) [reply cfg/DB/func/get_log.ddl:2]: its an direct output for front, I can change, but it will affect front

File: cfg/DB/table/logs.ddl (hunk 8 above)
Comment (SargtLa) [reply cfg/DB/table/logs.ddl:34]: its an refactor for old table, we need to change old logs to use with new features. there are new codes for send cv and interview. Earlier all these actions were recorded with code perform (as far as I remember)

File: api/avatar.go [hunk 9]
```diff
@@ -83,7 +83,7 @@ func HandleAddAvatar(ctx *fasthttp.RequestCtx) (interface{}, error) {
 	}
 
 	if i > 0 {
-		toLogCandidate(ctx, DB, id, "avatar", db.GetLogUpdateId())
+		toLogCandidateUpdate(ctx, DB, id, "{\"avatar\":\"\"}")
 	}
 
 	ctx.SetStatusCode(fasthttp.StatusAccepted)
```
Comment (ruslanBik4) [inline api/avatar.go]: Must be only if I >0

File: api/avatar.go [hunk 10]
```diff
@@ -83,7 +83,7 @@ func HandleAddAvatar(ctx *fasthttp.RequestCtx) (interface{}, error) {
 	}
 
 	if i > 0 {
-		toLogCandidate(ctx, DB, id, "avatar", db.GetLogUpdateId())
+		toLogCandidateUpdate(ctx, DB, id, "{\"avatar\":\"\"}")
```
Comment (ruslanBik4) [inline api/avatar.go]: I think that better define parameter as map[string] interface

Comment (ruslanBik4) [reply api/avatar.go]: I told about incapsulate DB into log interface yerstoday

File: api/to_log.go [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+	"github.com/ruslanBik4/dbEngine/dbEngine"
+	"github.com/ruslanBik4/logs"
+	"github.com/valyala/fasthttp"
+
+	"github.com/ruslanBik4/uppeople/auth"
+	"github.com/ruslanBik4/uppeople/db"
+)
+
+func HandleReturnLogsForCand(ctx *fasthttp.RequestCtx) (interface{}, error) {
+	DB, ok := ctx.UserValue("DB").(*dbEngine.DB)
```
Comment (ruslanBik4) [inline api/to_log.go:27]: We have here & above some code that get DB from ctx & invoke some its method.
Its' must be move into  a function

File: api/to_log.go [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+func HandleReturnLogsForCompany(ctx *fasthttp.RequestCtx) (interface{}, error) {
+	DB, ok := ctx.UserValue("DB").(*dbEngine.DB)
+	if !ok {
+		return nil, dbEngine.ErrDBNotFound
+	}
+
+	return DB.Conn.SelectToMaps(ctx, LOG_VIEW, ctx.UserValue("company_id"), false)
+}
+
+func toLogCandidateInsert(ctx *fasthttp.RequestCtx, DB *dbEngine.DB, candidateId int32, text string) {
```
Comment (ruslanBik4) [inline api/to_log.go]: use map instead string

Comment (ruslanBik4) [reply api/to_log.go]: Must be public & move into some module

File: api/to_log.go [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+			time.Now(),
+			code))
+}
+
+func toLogVacancy(ctx *fasthttp.RequestCtx, DB *dbEngine.DB, companyId, vacancyId int32, text string, code int32) {
+	toLog(ctx, DB,
+		dbEngine.ColumnsForSelect("user_id", "company_id", "vacancy_id", "text", "date_create",
+			"action_code"),
+		dbEngine.ArgsForSelect(auth.GetUserID(ctx), companyId, vacancyId,
+			text,
```
Comment (ruslanBik4) [inline api/to_log.go]: we may give map[string]interface parameters into db request

File: cfg/DB/func/get_log.ddl [hunk 14]
```diff
@@ -26,7 +26,44 @@ select logs.id as logId,
...
 
                   ELSE '' END,
-              CASE WHEN log_actions.is_insert_text = true THEN CONCAT(' ', logs.text) ELSE '' END
+              CASE WHEN log_actions.is_insert_text = true THEN
+                       CASE WHEN (log_actions.name = 'CODE_LOG_UPDATE'
+                           AND logs.text LIKE '{%}'
+                           AND logs.text::json is not null)
+                                THEN
+
+                                (SELECT DISTINCT array_to_string(array_agg(CONCAT(
```
Comment (ruslanBik4) [inline cfg/DB/func/get_log.ddl:35]: You don't used POstreSQL json function?

**Review by ruslanBik4 — CHANGES_REQUESTED** [review]

File: cfg/DB/func/get_log.ddl (hunk 14 above)
Comment (SargtLa) [reply cfg/DB/func/get_log.ddl:35]: I use some, it is a select from json_each()
But I do not use json_to_record, it don't meet my needs

File: api/candidate.go [hunk 15]
```diff
@@ -157,7 +157,7 @@ func HandleUpdateStatusCandidates(ctx *fasthttp.RequestCtx) (interface{}, error)
 
 	text := "новый статус кандидата по вакансии " + db.GetTagFromId(u.Status).Name
```
Comment (ruslanBik4) [inline api/candidate.go]: ? we need to call

File: api/avatar.go [hunk 16]
```diff
@@ -83,7 +83,7 @@ func HandleAddAvatar(ctx *fasthttp.RequestCtx) (interface{}, error) {
 	}
 
 	if i > 0 {
-		toLogCandidateUpdate(ctx, DB, id, "{\"avatar\":\"\"}")
+		toLogCandidateUpdate(ctx, id, map[string]interface{}{"avatar": "new"})
 	}
 
 	ctx.SetStatusCode(fasthttp.StatusAccepted)
```
Comment (ruslanBik4) [inline api/avatar.go]: mv up, after if I > 0

File: api/candidate.go (hunk 15 above)
Comment (ruslanBik4) [inline api/candidate.go]: db.GetTagFromId ? status != tag

File: api/candidate.go [hunk 17]
```diff
@@ -657,7 +657,7 @@ func HandleEditCandidate(ctx *fasthttp.RequestCtx) (interface{}, error) {
 	}
 
 	if i > 0 {
-		toLogCandidateUpdate(ctx, DB, id, loLogUpdateValues(columns, args))
+		toLogCandidateUpdate(ctx, id, loLogUpdateValues(columns, args))
 	}
 
 	ctx.SetStatusCode(fasthttp.StatusAccepted)
```
Comment (ruslanBik4) [inline api/candidate.go]: mv up

File: api/contacts.go [hunk 18]
```diff
@@ -83,7 +83,7 @@ func HandleAddContactForCompany(ctx *fasthttp.RequestCtx) (interface{}, error) {
 		}
 	}
 
-	toLogCompanyUpdate(ctx, DB, idCompany, " новый контакт "+u.Name)
+	toLogCompanyUpdate(ctx, idCompany, map[string]interface{}{"новый контакт": u.Name})
```
Comment (ruslanBik4) [inline api/contacts.go]: I think that we need a function toLogCompanyContact & save id contact there

**Review by ruslanBik4 — CHANGES_REQUESTED** [review]

File: api/avatar.go (hunk 16 above)
Comment (SargtLa) [reply api/avatar.go]: don't understand that

File: api/candidate.go (hunk 17 above)
Comment (SargtLa) [reply api/candidate.go]: don't understand that

**Review by ruslanBik4 — APPROVED** [review]

**Pull request merged**